kernel/arch/dreamcast/util/vmu_fb.c

```diff
@@ -0,0 +1,142 @@
+/* KallistiOS ##version##
+
+   util/vmu_fb.c
+   Copyright (C) 2023 Paul Cercueil
+
+*/
+
+#include <stdbool.h>
+#include <stdint.h>
+#include <string.h>
+
+#include <dc/maple/vmu.h>
+#include <dc/vmu_fb.h>
+
+#define GENMASK(h, l) \
+    (((unsigned int)-1 << (l)) & ((unsigned int)-1 >> (31 - (h))))
+
+static uint64_t extract_bits(const uint8_t *data,
+                             unsigned int offt, unsigned int w) {
+    uint32_t tmp, lsb, nb_bits;
+    uint64_t bits = 0;
+
+    /* This algorithm will extract "w" bits starting from bit "offt", and
+       place them right-adjusted in "bits".
+
+       Since we manipulate 8 bits at a time, and neither "w" nor "offt" are
+       required to be byte-aligned, we need to compute a mask of valid bits
+       for each byte that is processed. */
+    while (w) {
+        tmp = data[offt / 8];
+
+        if (8 - (offt & 0x7) > w)
+            lsb = 8 - (offt & 0x7) - w;
+        else
+            lsb = 0;
+
+        nb_bits = 8 - (offt & 0x7) - lsb;
+        bits <<= nb_bits;
+
+        tmp &= GENMASK(7 - (offt & 0x7), lsb);
+
+        bits |= tmp >> lsb;
+
+        offt += nb_bits;
+        w -= nb_bits;
+    }
+
+    return bits;
+}
+
+static void insert_bits(uint8_t *data,
+                        unsigned int offt, unsigned int w, uint64_t bits) {
+    uint32_t tmp, lsb, nb_bits, mask;
+
+    while (w) {
+        tmp = data[offt / 8];
+
+        if (8 - (offt & 0x7) > w)
+            lsb = 8 - (offt & 0x7) - w;
+        else
+            lsb = 0;
+
+        nb_bits = 8 - (offt & 0x7) - lsb;
+        mask = GENMASK(7 - (offt & 0x7), lsb);
+        tmp &= ~mask;
+
+        tmp |= ((bits >> (w - nb_bits)) << lsb) & mask;
+
+        data[offt / 8] = tmp;
+
+        offt += nb_bits;
+        w -= nb_bits;
+    }
+}
+
+void vmufb_paint_area(vmufb_t *fb,
+                      unsigned int x, unsigned int y,
+                      unsigned int w, unsigned int h,
+                      const char *data) {
+    unsigned int i;
+    uint64_t bits;
+
+    for (i = 0; i < h; i++) {
+        bits = extract_bits((const uint8_t *)data, i * w, w);
+        insert_bits((uint8_t *)fb->data, (y + i) * 48 + x, w, bits);
+    }
+}
+
+void vmufb_clear(vmufb_t *fb) {
+    memset(fb->data, 0, sizeof(fb->data));
+}
+
+void vmufb_clear_area(vmufb_t *fb,
+                      unsigned int x, unsigned int y,
+                      unsigned int w, unsigned int h) {
+    uint32_t tmp[48] = {};
+
+    vmufb_paint_area(fb, x, y, w, h, (const char *) tmp);
+}
+
+void vmufb_present(const vmufb_t *fb, maple_device_t *dev) {
```

Comment:
Oops. I screwed up and over-simplified it. You have to look at the connection direction of BOTH the peripheral AND the controller to know how to orient... VMU is always 0 (or connector facing up) which is the same as regular controllers, but something like the lightguns return 1 (connector facing down). I think this should generalize it:
`void vmufb_present(const vmufb_t *fb, maple_device_t *dev) {
    maple_device_t* cont = maple_enum_dev(dev->port, 0);

    /* Check for controller containing VMU (should always be same port, unit 0) */
    if(cont && (cont->info.functions & MAPLE_FUNC_CONTROLLER)) {
        /* If the VMU connector and controller connector face opposite directions, 
           no flipping necessary (VMU in a lightgun). */
        if(cont->info.connector_direction != dev->info.connector_direction) {
            vmu_draw_lcd(dev, fb->data);
            return;
        }
    }
    /* If we somehow found no corresponding controller, or connectors face the same direction,
       we must rotate the image 180 degreees (VMU in a standard controller). */
    vmu_draw_lcd_rotated(dev, fb->data);
}`

Reply:
Yep, no problem. I'm already in the VMU stuff adding date/time and button support. We're good!